### DISCUSSION TITLE
C/C++的整行读入

### DISCUSSION THREAD

**Discussion by SamuNatsu** [CSGrandeur/s-1problem1day1ac#417] [Knowledge]
# C/C++的整行读入
在C/C++中，使用`scanf("%s", str)`只能读取不带空格的字符串，有时候需要读入带空格的一整行字符，这时最简单粗暴的方法是`while((ch = getchar()) != '\n')`，逐字符写入数组，但这相略显麻烦。这里提供一些方法供大家参考。
## gets()
`gets()`可以从stdin中连续读取字符写入数组中直到出现换行符或文件尾，读取结束后，数组末尾的 `\n` 会被替换为 `\0` 。~~但`gets()`不检查边界，使用时需保证数组大小大于字符串长度，不安全，不推荐。~~ **gets()在C++11标准中已经被移除并使用gets_s()代替，考虑到市面上仍有部分OJ使用的是旧的C++98标准（如某万恶的POJ），最好不再使用gets()及gets_s()。**

## fgets()
函数原型：
```C++
char *fgets(char *str, int count, FILE *stream)
```
`fgets()`可以视为`gets()`的优化版。它多了两个参数，一个是stream，用于指定读取的文件流如stdin，另一个是count，指定最大写入字符数。与`gets()`不同的是，`fgets()`**会将读取到的**`\n`**保留在数组中，并在其后添加**`\0`。习惯上我们使用的数组的结尾是不带`\n` 的，因此使用`fgets()`读入的字符串**需要对末尾`\n`稍作处理**。

多组数据：
```C++
while(fgets(str, 1000000, stdin))
{
    str[strlen(str) - 1] = 0;
    \\ ...
}
```
## scanf()
其实 scanf() 也可以整行读入，只需：
```C++
scanf("%[^\n]%*c", str);
```
`scanf()`有两个比较少用的转换说明符`%[]`和`*`。`%[]`的功能与`%s`类似，但其只匹配`[]`中包含的字符，如`%[0-9]`为仅匹配数字，`%[a-zA-Z]`为仅匹配大小写字母。如果`[`后紧跟`^`则意为取补集，如`%[^0-9]`为匹配除数字以外的所有字符。`*`为抑制字符，跟在`%`后可以使当前转换说明符只匹配但不赋值给变量。

在上面的代码中，`%[^\n]`用于将`\n`之前的所有字符写入str中，`%*c`用于读取剩下来的`\n`但不赋给变量。

多组数据：
```C++
while(scanf("%[^\n]%*c", str) != EOF)
{
	...
}
```
## 注意事项
在整行读入前，需要先考虑上一行的`\n`是否残留并将其单独处理掉。
反面教材：
```C++
#include <bits/stdc++.h>
int a, b;
char s[100];
int main()
{
    scanf("%d%d", &a, &b);
    scanf("%[^\n]%*c", s); 
    printf("%s", s);
    return 0;
}
```
输入：
```
2 3
Hello world!
```
输出：

```
（无输出）
```
第一条`scanf()`语句读取后留下了行末的`\n`，被第二条`scanf()`读取到，导致s中没有写入任何内容。

在极少数情况下，OJ中的数据文件以`\r\n`作为每行的结尾，而上述各种方法都只会将`\r`视作普通字符，`fgets()`读取的字符串会以`\r\n`结尾，`scanf()`、`gets()`则会在结尾处会多出`\r`，都没有完美地处理这种情况。

不过，在明确OJ中以`\r\n`作为每行的结尾时，可以对`scanf()`稍加修改。
```C++
scanf("%[^\r]%*c%*c", str) \\此时读入以\n结尾的内容会出错
```
否则，在不确定OJ是否以`\r\n`结尾时，建议按`\n`的情况处理。如果追求鲁棒性，可以使用`getchar()`逐字符判断。
## std::getline()
在引入头文件`iostream`和`string`后，就可以使用一种省心又省力的行读取办法：
```cpp
std::string str;
std::getline(std::cin, str);
```
多组数据：
```cpp
std::string str;
while (std::getline(std::cin, str)) {
    ...
}
```
需要注意的是，一旦决定使用`std::cin`和`std::cout`来作为输入输出方式，那就需要考虑解除流同步和流绑定，因为`std::cin`和`std::cout`默认会与`scanf()`和`printf()`的缓冲区进行同步，以保证可以在代码中同时使用这四种输入输出方法而不出现乱序输入输出的情况，但这会大大拖慢`std::cin`和`std::cout`的速度，我们需要在程序开头加入这一段代码来解除同步和绑定：
```cpp
std::ios::sync_with_stdio(false);
std::cin.tie(0);
std::cout.tie(0);
```
在解除同步和绑定后，`std::cin`和`std::cout`就不能与`scanf()`和`printf()`同时使用了，但是换来了输入输出速度。

### 参考资料
[gets, gets_s - cppreference.com](https://en.cppreference.com/w/c/io/gets)
[fgets - cppreference.com](https://en.cppreference.com/w/c/io/fgets)
[scanf, fscanf, sscanf, scanf_s, fscanf_s, sscanf_s - cppreference.com](https://en.cppreference.com/w/c/io/fscanf)
[std::getline - cppreference.com](https://en.cppreference.com/w/cpp/string/basic_string/getline)

**Comment by 965295342** [comment]
yes
